apps/hash-ai-worker-ts/src/activities/infer-entities/update-entities.ts

```diff
@@ -0,0 +1,157 @@
+import type { VersionedUrl } from "@blockprotocol/type-system";
+import type { GraphApi } from "@local/hash-graph-client";
+import type {
+  InferredEntityUpdateFailure,
+  InferredEntityUpdateSuccess,
+} from "@local/hash-isomorphic-utils/temporal-types";
+import type {
+  AccountId,
+  Entity,
+  EntityId,
+  OwnedById,
+} from "@local/hash-subgraph";
+import {
+  extractEntityUuidFromEntityId,
+  extractOwnedByIdFromEntityId,
+} from "@local/hash-subgraph";
+
+import type { DereferencedEntityType } from "./dereference-entity-type";
+import { ProposedEntityUpdatesByType } from "./generate-tools";
+import { getEntityByFilter } from "./get-entity-by-filter";
+
+type StatusByTemporaryId<T> = Record<string, T>;
+
+type EntityUpdateStatusMap = {
+  updateSuccesses: StatusByTemporaryId<InferredEntityUpdateSuccess>;
+  updateFailures: StatusByTemporaryId<InferredEntityUpdateFailure>;
+};
+
+export const updateEntities = async ({
+  actorId,
+  graphApiClient,
+  log,
+  proposedEntityUpdatesByType,
+  requestedEntityTypes,
+}: {
+  actorId: AccountId;
+  graphApiClient: GraphApi;
+  log: (message: string) => void;
+  proposedEntityUpdatesByType: ProposedEntityUpdatesByType;
+  requestedEntityTypes: Record<
+    VersionedUrl,
+    { isLink: boolean; schema: DereferencedEntityType }
+  >;
+  ownedById: OwnedById;
+}): Promise<EntityUpdateStatusMap> => {
+  const entityStatusMap: EntityUpdateStatusMap = {
+    updateSuccesses: {},
+    updateFailures: {},
+  };
+
+  await Promise.all(
+    Object.values(requestedEntityTypes).map(async (entityType) => {
+      const entityTypeId = entityType.schema.$id;
+
+      const proposedUpdates = proposedEntityUpdatesByType[entityTypeId];
+
+      await Promise.all(
+        (proposedUpdates ?? []).map(async (proposedUpdate) => {
+          const { entityId, properties, updateEntityId } = proposedUpdate;
+
+          const proposedEntity = {
+            entityId,
+            properties,
+          };
+
+          let existingEntity: Entity | undefined = undefined;
+
+          try {
+            await graphApiClient.validateEntity(actorId, {
+              draft: true,
+              entityTypeId,
+              operations: ["all"],
+              properties,
+            });
+
```

Comment:
The `draft: true` in this means that the presence of required properties aren't checked, and since the entity already exists it should already have them – but there's no harm in doing as you suggest and it might avoid some other failure, so I've done this in https://github.com/hashintel/hash/pull/3659/commits/0583a05d2e7336b55e4c81bb3f49631636226526